crates/ingress-kafka/src/consumer_task.rs

```diff
@@ -457,36 +327,261 @@ impl Drop for AbortOnDrop {
+
+                    ingress_request_counter.increment(1);
+                    trace!(
+                        topic=%self.topic_partition.0,
+                        kafka_partition=%self.topic_partition.1,
+                        consumer_group=%self.consumer_group_id,
+                        offset=%offset,
+                        "store kafka offset",
+                    );
+
+                    self.consumer.store_offset(&self.topic_partition.0, self.topic_partition.1, offset)?;
+                },
+                Some(received) = consumer_stream.next() => {
+                    let msg = received?;
+                    let offset = msg.offset();
+                    if legacy_dedup_offset.is_some_and(|dedup_offset| offset as u64 <= dedup_offset) {
+                        // skip duplicated messages. Any gap should be small.
+                        debug!(
+                            topic=%self.topic_partition.0,
+                            kafka_partition=%self.topic_partition.1,
+                            consumer_group=%self.consumer_group_id,
+                            offset=%offset,
+                            "skipping kafka message (dedup)"
+                        );
+                        self.consumer.store_offset(&self.topic_partition.0, self.topic_partition.1, offset)?;
+                        continue;
+                    }
+
+                    trace!(
+                        topic=%self.topic_partition.0,
+                        kafka_partition=%self.topic_partition.1,
+                        consumer_group=%self.consumer_group_id,
+                        offset=%offset,
+                        "ingesting kafka message"
+                    );
+
+                    let envelope = self.builder.build(producer_id, &self.consumer_group_id, msg)?;
+
+                    let commit_token = self
+                        .ingestion
+                        .ingest(envelope.partition_key(), envelope)
+                        .await?
+                        .map(|_| offset);
+
+                    inflight.push_back(commit_token);
+                }
+            }
+        }
+    }
+
+    #[inline]
+    pub fn head_committed(
+        inflight: &mut VecDeque<RecordCommit<i64>>,
+    ) -> OptionFuture<&mut RecordCommit<i64>> {
+        OptionFuture::from(inflight.front_mut())
+    }
+}
+
+// Do not change. Changing this hasher will create new producer-id which can
+// cause duplicates
+fn dedup_producer_id(
+    subscription: &SubscriptionId,
+    consumer_group: &str,
+    topic: &str,
+    partition: i32,
+) -> u128 {
+    let mut hasher = xxhash_rust::xxh3::Xxh3::new();
+
+    subscription.hash(&mut hasher);
+    '\0'.hash(&mut hasher);
+    consumer_group.hash(&mut hasher);
+    '\0'.hash(&mut hasher);
+    topic.hash(&mut hasher);
+    '\0'.hash(&mut hasher);
+    partition.hash(&mut hasher);
+
+    hasher.digest128()
 }
```

Comment:
Wondering whether u128 is actually an overkill for the `ProducerId::Producer`. How many producers could there be and could we generate the Kafka producer id based on the Schema (e.g. having a counter value that counts the number of subscriptions and every new subscription gets the next counter value)? If we wouldn't hash then decreasing the producer id value to a u64 should be more than sufficient. Of course, one would have to handle the different partitions which could be done by using a few bits of the u64 for the partition (I guess that 16 bist should be more than enough for the number of partitions).

Nothing we have to address right now but maybe something to double check before creating the v1.6 release.